File: 2024/day03/clojure/solution.clj
(ns solution)

(defn part1
  "Find all valid mul(X,Y) instructions and sum their products."
  [data]
  (let [pattern #"mul\((\d{1,3}),(\d{1,3})\)"
        matches (re-seq pattern data)]
    (->> matches
         (map (fn [[_ x y]]
                (* (parse-long x) (parse-long y))))
         (reduce + 0))))

(defn part2
  "Like part1, but do() enables and don't() disables mul instructions."
  [data]
  (let [mul-pattern #"mul\((\d{1,3}),(\d{1,3})\)"
        do-pattern #"do\(\)"
        dont-pattern #"don't\(\)"

        ;; Use re-matcher to get all matches with positions
        events (concat
                ;; Get all mul instructions with positions
                (loop [matcher (re-matcher mul-pattern data)
                       results []]
                  (if (.find matcher)
                    (recur matcher
                           (conj results {:pos (.start matcher)
                                         :type :mul
                                         :x (parse-long (.group matcher 1))
                                         :y (parse-long (.group matcher 2))}))
                    results))

                ;; Get all do() instructions with positions
                (loop [matcher (re-matcher do-pattern data)
                       results []]
                  (if (.find matcher)
                    (recur matcher
                           (conj results {:pos (.start matcher)
                                         :type :do}))
                    results))

                ;; Get all don't() instructions with positions
                (loop [matcher (re-matcher dont-pattern data)
                       results []]
                  (if (.find matcher)
                    (recur matcher
                           (conj results {:pos (.start matcher)
                                         :type :dont}))
                    results)))

        ;; Sort events by position
        sorted-events (sort-by :pos events)]

    ;; Process events in order
    (loop [events sorted-events
           enabled true
           total 0]
      (if (empty? events)
        total
        (let [event (first events)]
          (case (:type event)
            :do (recur (rest events) true total)
            :dont (recur (rest events) false total)
            :mul (if enabled
                   (recur (rest events) enabled (+ total (* (:x event) (:y event))))
                   (recur (rest events) enabled total))))))))

(defn -main []
  (let [;; Read from ../input.txt relative to current working directory
        input-path "../input.txt"
        data (slurp input-path)]
    (println "Part 1:" (part1 data))
    (println "Part 2:" (part2 data))))

(-main)

File: 2024/day03/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use File::Spec;

sub part1 {
    my ($data) = @_;
    my $total = 0;

    # Find all valid mul(X,Y) instructions where X and Y are 1-3 digit numbers
    while ($data =~ /mul\((\d{1,3}),(\d{1,3})\)/g) {
        $total += $1 * $2;
    }

    return $total;
}

sub part2 {
    my ($data) = @_;
    my $total = 0;
    my $enabled = 1;

    # Build a list of all events with positions
    my @events;

    # Find all mul instructions
    while ($data =~ /mul\((\d{1,3}),(\d{1,3})\)/g) {
        push @events, [$-[0], 'mul', $1, $2];
    }

    # Find all do() instructions
    while ($data =~ /do\(\)/g) {
        push @events, [$-[0], 'do', 0, 0];
    }

    # Find all don't() instructions
    while ($data =~ /don't\(\)/g) {
        push @events, [$-[0], 'dont', 0, 0];
    }

    # Sort by position
    @events = sort { $a->[0] <=> $b->[0] } @events;

    # Process events in order
    for my $event (@events) {
        my ($pos, $type, $x, $y) = @$event;

        if ($type eq 'do') {
            $enabled = 1;
        } elsif ($type eq 'dont') {
            $enabled = 0;
        } elsif ($type eq 'mul' && $enabled) {
            $total += $x * $y;
        }
    }

    return $total;
}

sub main {
    # Get the directory of the script
    my $script_dir = dirname(__FILE__);
    my $input_path = File::Spec->catfile($script_dir, '..', 'input.txt');

    # Read the input file
    open my $fh, '<', $input_path or die "Cannot open $input_path: $!";
    my $data = do { local $/; <$fh> };
    close $fh;

    print "Part 1: ", part1($data), "\n";
    print "Part 2: ", part2($data), "\n";
}

main();
